web/src/message_store.ts

```diff
@@ -1,83 +1,110 @@
 import _ from "lodash";
+import {z} from "zod";
 
 import * as blueslip from "./blueslip";
 import * as people from "./people";
-import type {Submessage} from "./submessage";
-import type {TopicLink} from "./types";
+import {topic_link_schema} from "./types";
 import type {UserStatusEmojiInfo} from "./user_status";
 
 const stored_messages = new Map<number, Message>();
 
-export type MatchedMessage = {
-    match_content?: string | undefined;
-    match_subject?: string | undefined;
-};
-
-export type MessageReactionType = "unicode_emoji" | "realm_emoji" | "zulip_extra_emoji";
-
-export type DisplayRecipientUser = {
-    email: string;
-    full_name: string;
-    id: number;
-    is_mirror_dummy: boolean;
-    unknown_local_echo_user?: boolean;
-};
-
-export type DisplayRecipient = string | DisplayRecipientUser[];
-
-export type MessageEditHistoryEntry = {
-    user_id: number | null;
-    timestamp: number;
-    prev_content?: string;
-    prev_rendered_content?: string;
-    prev_rendered_content_version?: number;
-    prev_stream?: number;
-    prev_topic?: string;
-    stream?: number;
-    topic?: string;
-};
-
-export type MessageReaction = {
-    emoji_name: string;
-    emoji_code: string;
-    reaction_type: MessageReactionType;
-    user_id: number;
-};
-
-export type RawMessage = {
-    avatar_url: string | null;
-    client: string;
-    content: string;
-    content_type: "text/html";
-    display_recipient: DisplayRecipient;
-    edit_history?: MessageEditHistoryEntry[];
-    id: number;
-    is_me_message: boolean;
-    last_edit_timestamp?: number;
-    reactions: MessageReaction[];
-    recipient_id: number;
-    sender_email: string;
-    sender_full_name: string;
-    sender_id: number;
-    sender_realm_str: string;
-    submessages: Submessage[];
-    timestamp: number;
-    flags: string[];
-} & (
-    | {
-          type: "private";
-      }
-    | {
-          type: "stream";
-          stream_id: number;
-          // Messages that come from the server use `subject`.
-          // Messages that come from `send_message` use `topic`.
-          subject?: string;
-          topic?: string;
-          topic_links: TopicLink[];
-      }
-) &
-    MatchedMessage;
+const matched_message_schema = z.object({
+    match_content: z.optional(z.string()),
+    match_subject: z.optional(z.string()),
+});
+
+export type MatchedMessage = z.infer<typeof matched_message_schema>;
+
+const message_reaction_type_schema = z.enum(["unicode_emoji", "realm_emoji", "zulip_extra_emoji"]);
+
+export type MessageReactionType = z.infer<typeof message_reaction_type_schema>;
+
+const display_recipient_user_schema = z.object({
+    email: z.string(),
+    full_name: z.string(),
+    id: z.number(),
+    is_mirror_dummy: z.boolean(),
+    unknown_local_echo_user: z.optional(z.boolean()),
```

Comment:
This is not part of the API; it's something that `echo.ts` adds to objects. Probably it's not ideal for that to be part of the `RawMessage` type.

Though looking at this, I think we might be able to rip out the section of `build_display_recipient` that uses `unknown_local_echo_user`, since I think the pills transition for the compose direct message recipient component, which now stores users using user IDs exclusively, removed support for just typing in an `example@mit.edu` email address in the Zulip compose area and sending a Zephyr to them despite them not existing in the Zulip system. @andersk does that sound correct?

If so, then the right answer is a prep commit to remove the code found via `git grep unknown_local_echo_user`, and then this type can accurately match the API definitions.

Reply:
If we have decided to remove `unknown_local_user_echo` definitively , I can remove it from `server_message` schema and `build_display_recipient` function since I am migrating `echo.js` to TypeScript.

- https://github.com/zulip/zulip/pull/30702

Reply:
Should I drop this commit here?

Reply:
I think we can keep it. Depending on which PR happens to get merged first, we can rebase the other one.

Reply:
- https://github.com/zulip/zulip/pull/30702/commits/bb7c8742136be24d41d8cea502a4099f98e94e5e